pkg/controller/kubeconfig.go

```diff
@@ -6,60 +6,129 @@ import (
+		return nil, err
+	}
+
+	outConfig := infoKubeconfig.DeepCopy()
+
+	outConfig.AuthInfos = map[string]*clientcmdapi.AuthInfo{
+		"": {
+			Token: token,
+		},
+	}
+
+	return outConfig, nil
+}
+
 func (c *Controller) createBootstrapToken(name string) (string, error) {
+	existingSecret, err := c.getSecretIfExists(name)
+	if err != nil {
+		return "", err
+	}
+	if existingSecret != nil {
+		return c.updateSecretExpirationAndGetToken(existingSecret)
+	}
+
 	tokenID := rand.String(6)
 	tokenSecret := rand.String(16)
 
 	secret := v1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: fmt.Sprintf("bootstrap-token-%s", tokenID),
+			Name:   fmt.Sprintf("bootstrap-token-%s", tokenID),
+			Labels: map[string]string{machineNameLabelKey: name},
 		},
 		Type: secretTypeBootstrapToken,
 		StringData: map[string]string{
 			"description":                    "bootstrap token for " + name,
-			"token-id":                       tokenID,
-			"token-secret":                   tokenSecret,
-			"expiration":                     metav1.Now().Add(24 * time.Hour).Format(time.RFC3339),
+			tokenIDKey:                       tokenID,
+			tokenSecretKey:                   tokenSecret,
+			expirationKey:                    metav1.Now().Add(24 * time.Hour).Format(time.RFC3339),
 			"usage-bootstrap-authentication": "true",
 			"usage-bootstrap-signing":        "true",
 			"auth-extra-groups":              "system:bootstrappers:machine-controller:default-node-token",
 		},
 	}
 
-	_, err := c.kubeClient.CoreV1().Secrets(metav1.NamespaceSystem).Create(&secret)
+	_, err = c.kubeClient.CoreV1().Secrets(metav1.NamespaceSystem).Create(&secret)
 	if err != nil {
 		return "", err
 	}
 
-	return fmt.Sprintf("%s.%s", tokenID, tokenSecret), nil
+	return fmt.Sprintf(tokenFormatter, tokenID, tokenSecret), nil
 }
 
-func (c *Controller) createBootstrapKubeconfig(name string) (*clientcmdapi.Config, error) {
-	token, err := c.createBootstrapToken(name)
+func (c *Controller) updateSecretExpirationAndGetToken(secret *v1.Secret) (string, error) {
+	if secret.StringData == nil {
+		secret.StringData = map[string]string{}
+	}
+	secret.StringData[expirationKey] = metav1.Now().Add(24 * time.Hour).Format(time.RFC3339)
+	tokenID := secret.StringData[tokenIDKey]
+	tokenSecret := secret.StringData[tokenSecretKey]
+	token := fmt.Sprintf(tokenFormatter, tokenID, tokenSecret)
+
+	expBytes, ok := secret.Data["expiration"]
+	if !ok {
+		return "", fmt.Errorf("haven't found %s key in the secret's Data field", expirationKey)
+	}
+	expString := string(expBytes)
+	expVal := metav1.Now()
+	err := expVal.UnmarshalQueryParameter(expString)
 	if err != nil {
-		return nil, err
+		return "", err
+	}
+	now := metav1.Now()
+	// expVal has to point to a time in the future otherwise we need to update expiration time
+	if now.Before(&expVal) {
+		return token, nil
 	}
 
-	infoKubeconfig, err := c.kubeconfigProvider.GetKubeconfig()
+	_, err = c.kubeClient.CoreV1().Secrets(metav1.NamespaceSystem).Update(secret)
```

Comment:
haven't we made this token immortal :( ?

Reply:
I would remove "extend token expiration" logic. If the token is no longer valid it will be `GCed` and we will crate a new valid token for a machine. WDYT ?

Reply:
makes sense, then we should add some duration to `now`, so that there is enough time for a machine to join the cluster (`now.Before(&expVal`)

Reply:
Yes. We need some definition. 30min? Wdyt?

Reply:
I set it to 15 min :)